controlplane/kubeadm/internal/machinefilters/machine_filters.go

```diff
@@ -277,8 +277,14 @@ func matchClusterConfiguration(kcp *controlplanev1.KubeadmControlPlane, machine
 	if ok {
 		machineClusterConfig := &kubeadmv1.ClusterConfiguration{}
 		// ClusterConfiguration annotation is not correct, only solution is to rollout.
-		if err := json.Unmarshal([]byte(machineClusterConfigStr), machineClusterConfig); err != nil {
-			return false
+		// NOTE: it is required to handle a nil ClusterConfiguration (that is serialized into "null") as a special case,
+		// because Unmarshal returns &kubeadmv1.ClusterConfiguration{}, not nil; see https://github.com/kubernetes-sigs/cluster-api/issues/3353
```

Comment:
+1 to ensuring that we handle this since we've already shipped a release with the current behavior, but should we also avoid serializing the clusterConfiguration to null when we create the annotation as well?

Reply:
Uhhm I'm not sure this is correct because otherwise there is no way to distinguish ClusterConfiguration=nil from the absence of annotation, and the latter is managed as "We don't have enough information to make a decision; don't' trigger a roll out" (while the former should trigger a comparison).

Reply:
I'm wondering if there is some other way that we can convey the absence of the annotation? Otherwise we would trigger a rollout if for some reason ClusterConfiguration goes from nil to ClusterConfiguration{}, both of which essential mean default for our purposes.

Not necessarily something we need to solve now, but feels like something that will eventually come back to bite us eventually.

Reply:
@detiber What about treating a nil or empty ClusterConfiguration the same way?

Reply:
@vincepri +1 sgtm